### DISCUSSION TITLE
Excessive size of maturin-pypi releases

### DISCUSSION THREAD

**Discussion by lukaslueg** [PyO3/maturin#1962] [Q&A] [answered]
I recently pushed a maturin-built project to pypi, mostly for fun profit, using maturin's own github action. I was surprised to find that a point release built and deployed by maturin's default github action consumes around 20mb of space on pypi, which clocks in against pypi's default upload limit of 100mb.

Maturin builds 65 wheels by default, each of which comes in at around 300kb (yielding around 20mb in total) for me. Having configured for stripped lto builds, most (around 250kb) of the binary is consumed by stdlib. It seems unfeasible to reduce the size of the binary further.

Am I missing something obvious here? My impression is that building 65 wheels in a default configuration, which immediately consumes one fives of pypi's upload limit, for a project which brings in around 50kb of own code, seems rather excessive.

**Comment by konstin** [comment]
This sounds like way more wheels than i'd expect, are you using the `abi3` feature? You can also save by not uploading for rarer architectures and only provide wheels for the main ones (x86_64/aarch64 glibc linux, x86_64 windows and x86_64/aarch64 mac)

**Reply by lukaslueg** [reply]
No, I'm not using `abi3` in the manifest; the workflow is also [pretty much](https://github.com/lukaslueg/pyhyperminhash/blob/master/.github/workflows/CI.yml) what maturin provides by default.

I was hesitant to restrict the Python compatibility, because the package itself (when built from src) should be compatible everywhere; I only [restricted](https://github.com/lukaslueg/pyhyperminhash/blob/master/pyproject.toml) `<3.13` due to #1960. While I don't see much gain in building wheels for `cpy38` and `cpy39`, I fail to find a knob to turn that off.

**Reply by konstin** [reply, marked as answer]
I would highly recommend trying the `abi3` feature, this will reduce the number of builds to one per target triple instead of one per target triples and python version.

**Reply by lukaslueg** [reply]
Thanks for pointing this out.
